### PR TITLE
Adding dcterms:description and dcterms:title property shapes where the valueType is String instead of XmlLiteral.

### PR DESCRIPTION
**Pull request by jadelkhoury**
## Description

Adding dcterms:description and dcterms:title property shapes where the valueType is String instead of XmlLiteral.
Note that the main change is in the model. The model changes do not result in code changes.
The code changes are due to generation baed on latest LyoDesigner.

## Checklist

- [ ] This PR adds an entry to the CHANGELOG. _See https://keepachangelog.com/en/1.0.0/ for instructions. Minor edits are exempt._
- [X] This PR was tested on at least one Lyo OSLC server or adds unit/integration tests.
- [X] This PR does NOT break the API

### CODE REVIEW

**Comment by jadelkhoury** [comment]
@berezovskyi @jamsden  It would be interesting to understand the motivation to have XmlLiteral properties for things like title and description.

**Comment by jamsden** [comment]
To support formatted text (XHTML) in description properties.

**Comment by jadelkhoury** [comment]
The problem occurs when Description is **not** HTML, but a simple string. Then things like "&" in the string cause problems.
THe proposal here in Lyo.Domains is to support both XmlLIteral and String.

**Comment by berezovskyi** [comment]
To be clear, problems appear even when the value is valid HTML but not XML. With the near-abandonment of XHTML 1.0 in favor of HTML 5, this is a major design deficiency in the choice of value type. I think a bnode like

```
:r1 dcterms:description [ a oslc:HtmlFragment;
                          rdf:value "<strong>REQ-15</strong> The property shall access valid HTML 5 fragment values"
                        ].
```

would have been the right choice. BTW, `rdf:value` is a real property AND is specifically designed for such use-case: https://www.w3.org/TR/rdf-schema/#ch_value

**Comment by berezovskyi** [comment]
@jad-elkhoury can I merge master into your branch to fix the build?

**Comment by jadelkhoury** [comment]
Sure, if that fixes the build. 

And I agree with you @berezovskyi . I was working with html stuff, but that was also problematic.

**Comment by jad-elkhoury** [comment]
any update on this pull request? OK to merge?

**Comment by jamsden** [comment]
We can't change the data type of dcterms:description in a way that would be a breaking change. A lot of code expects this to be an XMLLiteral.

**Comment by jadelkhoury** [comment]
@jamsden! we are not changing. I am adding a new property constraint with a String datatype. and both constraints will still point to the dcterms:description vocabulary term.

Our OSLC domains (RM, QM) are still be using the expected dcterms:description that is XMLLiteral.

File: domains/oslc-domains/pom.xml [hunk 1]
```diff
@@ -16,20 +16,45 @@
     -->
     <groupId>org.eclipse.lyo</groupId>
     <artifactId>oslc-domains</artifactId>
-    <version>4.2.0-SNAPSHOT</version>
+    <version>4.0.0</version>
```
Comment (berezovskyi) [inline domains/oslc-domains/pom.xml:19]: All the POM changes need to be rolled back.

File: domains/oslc-domains/pom.xml [hunk 2]
```diff
@@ -42,23 +67,30 @@
...
+            <version>3.1.0</version>
             <scope>provided</scope>
         </dependency>
         <dependency>
-            <groupId>javax.servlet.jsp.jstl</groupId>
-            <artifactId>javax.servlet.jsp.jstl-api</artifactId>
-            <scope>provided</scope>
+            <groupId>javax.servlet</groupId>
+            <artifactId>jstl</artifactId>
+            <version>1.2</version>
```
Comment (berezovskyi) [inline domains/oslc-domains/pom.xml:86]: @jadelkhoury maybe you should also change LyoD to use the artifact on the left for the JSTL.

File: domains/org.eclipse.lyo.tools.domainmodels/oslcDomainSpecifications.xml [hunk 3]
```diff
@@ -13,6 +13,9 @@
     <resourceProperties id="_pnXOMpUdEeq-KoPaR9_Cfg" name="description" description="Descriptive text about resource represented as rich text in XHTML content. SHOULD include only content that is valid and suitable inside an XHTML &lt;div> element." occurs="zeroOrOne" valueType="XMLLiteral">
       <propertyDefinition href="vocabulary.xml#//@vocabularies[label='Dublin%20Core']/@properties[name='description']"/>
     </resourceProperties>
+    <resourceProperties id="_fdNWAFv1EeyHYL9SbPncDg" name="description" description="Descriptive text about resource represented as rich text in XHTML content. SHOULD include only content that is valid and suitable inside an XHTML &lt;div> element." occurs="zeroOrOne" valueType="String">
```
Comment (berezovskyi) [inline domains/org.eclipse.lyo.tools.domainmodels/oslcDomainSpecifications.xml:16]: @jamsden haha here is where the weasel-wording bit us! Jena and Java do not care about our SHOULD (be valid XHTML) and throw exceptions quite happily if you feed them with bad XML.

**Review by berezovskyi — CHANGES_REQUESTED** [review]
Jad, thanks. The POM file changes need to be rolled back, please. Did you actually re-generate the code? I didn't see any changes except for the labels.

**Comment by berezovskyi** [comment]
> We can't change the data type of dcterms:description in a way that would be a breaking change. A lot of code expects this to be an XMLLiteral.

Jim, we can do whatever we want in Lyo to maintain compatibility. Lyo is not a spec, though we need to bump the major version if the Java interface compatibility breaks (guess what, it doesn't because we were backing `ValueType.XMLLiteral` properties with `String` fields already; that code was written long before me, so project leads till 2013 knew about invalid XMLLiteral strings flying around). We had multiple reports of Jazz returning bad XML in the properties that were supposed to be `XMLLiteral`. Those users need something to fix the problem. Jena stores XMLLiteral as a String anyway without additional ceremony (other than the XML validity check that many strings in the wild fail). On top of that, JenaModelHelper includes a workaround that looks similar to this in pseudocode:

```
if (valuetype == XMLLiteral) {
  try {
    parseXMLLiteral()
  } exception (ScaryXmlException e) {
	setValue(value.toString()) 
  }
}
```

The problem is that this workaround is only applied when:

1. A property is declared.
2. The property `ValueType` is `ValueType.XMLLiteral` but the field type is `String`.

As users on OSLC and Jazz forums discovered quite painfully, this workaround only kicks in on defined shape properties and does not apply in "extended" properties. We are planning to let users choose in LyoD whether to back a `ValueType.XMLLiteral` with a `java.lang.String` or a `org.eclipse.lyo.oslc4j.core.model.XMLLiteral`, so that the workarounds are disabled for the latter and tell users that the tool they are interfacing is sending them 0x0badf00d.

**Comment by berezovskyi** [comment]
Jena link: https://github.com/apache/jena/blob/56534eac635c810464bc99351baa69c728a9dd58/jena-core/src/main/java/org/apache/jena/datatypes/xsd/impl/XMLLiteralType.java#L40

**Comment by berezovskyi** [comment]
@jadelkhoury how did you build this? Maven is complaining:

>  Error:  Failed to execute goal org.apache.maven.plugins:maven-enforcer-plugin:3.0.0:enforce (enforce) on project lyo-parent: Some Enforcer rules have failed

Did I tell you how much I love `maven-enforcer-plugin`, by the way?

File: domains/oslc-domains/pom.xml (hunk 1 above)
Comment (jadelkhoury) [reply domains/oslc-domains/pom.xml:19]: Can you clarify? why do we need to roll back?

File: domains/oslc-domains/pom.xml (hunk 2 above)
Comment (jadelkhoury) [reply domains/oslc-domains/pom.xml:86]: ???

**Comment by jadelkhoury** [comment]
Could it be that I made this commit before we released 4.1.0? that's why it fails in the build now?

**Comment by berezovskyi** [comment]
No, you messed with the POM file a bit and my checks failed the build :)

File: domains/org.eclipse.lyo.tools.domainmodels/oslcDomainSpecifications.xml [hunk 4]
```diff
@@ -82,18 +88,21 @@
...
     <resourceProperties id="_posD4JUdEeq-KoPaR9_Cfg" name="comment" description="May be used to provide a human-readable description of a resource." valueType="XMLLiteral">
       <propertyDefinition href="vocabulary.xml#//@vocabularies[label='RDFS']/@properties[name='comment']"/>
     </resourceProperties>
+    <resourceProperties id="_rZmPgFv1EeyHYL9SbPncDg" name="comment" description="May be used to provide a human-readable description of a resource." valueType="String">
+      <propertyDefinition href="vocabulary.xml#//@vocabularies[label='RDFS']/@properties[name='comment']"/>
+    </resourceProperties>
     <resourceProperties id="_poyxkpUdEeq-KoPaR9_Cfg" name="seeAlso" description="Used to indicate a resource that might provide additional information about the subject resource." occurs="zeroOrOne" valueType="URI">
       <propertyDefinition href="vocabulary.xml#//@vocabularies[label='RDFS']/@properties[name='seeAlso']"/>
     </resourceProperties>
-    <resourceProperties id="_uuAX4JUdEeq-KoPaR9_Cfg" name="label" valueType="String">
```
Comment (jad-elkhoury) [inline domains/org.eclipse.lyo.tools.domainmodels/oslcDomainSpecifications.xml:94]: Note that this property was not deleted, It still exists (line 91), including a description of the property.

File: domains/org.eclipse.lyo.tools.domainmodels/oslcDomainSpecifications.xml [hunk 5]
```diff
@@ -82,18 +88,21 @@
     <resourceProperties id="_posD4ZUdEeq-KoPaR9_Cfg" name="label" description="May be used to provide a human-readable version of a resource's name." occurs="zeroOrOne" valueType="XMLLiteral">
       <propertyDefinition href="vocabulary.xml#//@vocabularies[label='RDFS']/@properties[name='label']"/>
     </resourceProperties>
+    <resourceProperties id="_uuAX4JUdEeq-KoPaR9_Cfg" name="label" description="May be used to provide a human-readable version of a resource's name." occurs="zeroOrOne" valueType="String">
```
Comment (jad-elkhoury) [inline domains/org.eclipse.lyo.tools.domainmodels/oslcDomainSpecifications.xml:91]: Note that this property was not added, It was a modification of what we have on line 94 in old version. We include a description of the property. We also moved the definiition to be next to the "label" that is of type XmlLiteral.

**Comment by jad-elkhoury** [comment]
@berezovskyi I made the requested changes. We good to go?

**Review by berezovskyi — APPROVED** [review]

**Pull request merged**